content/en/blog/_posts/clientcmd-tutorial.md

```diff
@@ -0,0 +1,275 @@
+you’ll see how to do this [later on](#bind-the-flags).
+
+## Available features
+
+`clientcmd` allows programs to handle
+
+* `kubeconfig` selection (using `KUBECONFIG`);
+* context selection;
+* namespace selection;
+* client certificates;
+* user impersonation;
+* basic authentication support (username/password).
+
+## Configuration merging
+
+In various scenarios, `clientcmd` supports _merging_ configuration settings:
+`KUBECONFIG` can specify multiple files whose contents are combined.
+This can be confusing, because settings are merged in different directions
+depending on how they are implemented.
+If a setting is defined in a map, the first definition wins,
+subsequent definitions are ignored.
+If a setting is not defined in a map, the last definition wins.
+
+When settings are retrieved using `KUBECONFIG`,
+missing files result in warnings only.
+If the user explicitly specifies a path (in `--kubeconfig` style),
+there must be a corresponding file.
+
+If `KUBECONFIG` isn’t defined,
+the default configuration file, `~/.kube/config`, is used instead,
+if present.
+
+### Overall process
+
+The general usage pattern is succinctly expressed in
+[the package documentation](https://pkg.go.dev/k8s.io/client-go/tools/clientcmd):
+
+```lang-go
+loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+// if you want to change the loading rules (which files in which order), you can do so here
+
+configOverrides := &clientcmd.ConfigOverrides{}
+// if you want to change override values or bind them to flags, there are methods to help you
+
+kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
+config, err := kubeConfig.ClientConfig()
+if err != nil {
+	// Do something
+}
+client, err := metav1.New(config)
+// ...
+```
+
+In the context of this article, there are six steps:
+
+1. Configure the loading rules.
+1. Configure the overrides.
+1. Build a set of flags.
+1. Bind the flags.
+1. Build the merged configuration.
+1. Obtain an API client.
+
+### Configure the loading rules
+
+`clientcmd.NewDefaultClientConfigLoadingRules()` builds loading rules which will use either the contents of the `KUBECONFIG` environment variable,
+or the default configuration file name (`~/.kube/config`).
+In addition, if the default configuration file is used,
+it is able to migrate settings from the (very) old default configuration file
+(`~/.kube/.kubeconfig`).
+
+You can build your own `ClientConfigLoadingRules`,
+but in most cases the defaults are fine.
+
+### Configure the overrides
+
+`clientcmd.ConfigOverrides` is a `struct` storing overrides which will be applied over the settings loaded from the configuration derived using the loading rules.
+In the context of this article,
+its primary purpose is to store values obtained from command-line flags.
+
+In most cases there’s nothing to set in the overrides;
+we will only bind them to flags.
+
+### Build a set of flags
```

Comment:
Although Go calls these things flags, the wider world of IT uses “flag“ most commonly to mean a boolean. See https://en.wikipedia.org/wiki/Flag_(programming) for more context.

For example, `--enable-foo` is a flag but `--foo-path=/tmp` is not.

To help readers, especially those new to Kubernetes or less experienced in Go, call them command line arguments. You can also use _flag_ when you are talking about the Go library code specifically.

Reply:
That’s a good point, thanks! I’ve reworked the sections discussing flags, to avoid using “flag” where it should be “command line argument”, and to clarify what is meant by “flag” (really, the `FlagInfo` `struct` — that explains the `clientcmd` function names ending with `Flags`). I’ve also dropped the “command-line” hyphen, and a mention of “options” instead of “arguments”.